### DISCUSSION TITLE
Use a Custom Allocator for Parse Tree Nodes to Improve Parsing Performance

### DISCUSSION THREAD

**Discussion by nordlow** [tree-sitter/tree-sitter#3943] [Ideas]
### Feature Request

Has anyone explored the possibility of accelerating parsing by employing a region allocator for all nodes in a parse tree? This approach would enhance performance in cases where a single parse tree is constructed in one go, without subsequent edits. I'm not sure whether it could be utilized in the incremental reparsing case aswell.

The idea is to store all allocations associated with a parse tree consecutively in memory, potentially reducing memory fragmentation and improving cache locality and of course speed of a single allocation being a single pointer bump.

For instance, https://memory-pool-system.readthedocs.io/en/latest/ could be used. Note that MPS is currently planned to be used in Emacs to speed up its GC. See Emacs branch `scratch/igc` for details.

I am aware of that the current C-style memory management used by tree-sitter can be overridden.

**Comment by amaanq** [comment]
In general, for open-ended questions like this it'd be better to start a discussion rather than open an issue
